File: Codeforces/2238/C/try-1.py
/*
 * Problem : Village Guilds
 * Difficulty: Medium
 * Submission: Try 1
 * status: Accepted
 * Language: Python 3
 * Date: 7/7/2026, 7:47:59 PM
 * Link: https://codeforces.com/contest/2238/problem/C
 */

import sys

sys.setrecursionlimit(1_000_000)
input = sys.stdin.readline

def dfs(v, p):
    global ans, dp, depth, g

    ans[v] = 0
    dp[v] = depth[v]

    m1 = depth[v]
    m2 = depth[v]

    for u in g[v]:
        if u != p:
            depth[u] = depth[v] + 1
            dfs(u, v)

            dp[v] = max(dp[v], dp[u])
            ans[v] += ans[u]

            if dp[u] >= m1:
                m2 = m1
                m1 = dp[u]
            elif dp[u] >= m2:
                m2 = dp[u]

    ans[v] += m2 - depth[v] + 1


t = int(input())

for _ in range(t):
    n = int(input())

    g = [[] for _ in range(n)]
    depth = [0] * n
    dp = [0] * n
    ans = [0] * n

    if n > 1:
        parents = list(map(int, input().split()))
        for i, p in enumerate(parents, start=1):
            g[p - 1].append(i)

    dfs(0, -1)
    print(ans[0])

File: LeetCode/Medium/3532-path-existence-queries-in-a-graph-i/try-1.java
/*
 * Problem #3532: Path Existence Queries in a Graph I
 * Difficulty: Medium
 * Submission: Try 1
 * status: Accepted
 * Language: java
 * Date: 7/9/2026, 7:44:10 PM
 * Link: https://leetcode.com/problems/path-existence-queries-in-a-graph-i/
 */

class Solution {
    public boolean[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {

        int[] comp = new int[n];
        int id = 0;

        for (int i = 1; i < n; i++) {
            if (nums[i] - nums[i - 1] > maxDiff)
                id++;
            comp[i] = id;
        }

        boolean[] ans = new boolean[queries.length];

        for (int i = 0; i < queries.length; i++) {
            ans[i] = comp[queries[i][0]] == comp[queries[i][1]];
        }

        return ans;
    }
}
